### PR TITLE
LCSD-4851 Tied house exemption Invitation

### PR DESCRIPTION
**Pull request by GeorgeWalker**

### CODE REVIEW

File: cllc-public-app/ViewModels/ApplicationTypeNames.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+using System;
+using System.Runtime.Serialization;
+using System.Text.Json.Serialization;
+using Newtonsoft.Json.Converters;
+
+namespace Gov.Lclb.Cllb.Public.ViewModels
+{
+    public static class ApplicationTypeNames
+    {
+        public const string TiedHouseExcemption = "Tied House Exemption Application";
```
Comment (asanchezr) [inline cllc-public-app/ViewModels/ApplicationTypeNames.cs:10]: very small typo `TiedHouseEx(c)emption` -> `TiedHouseExemption`

File: carla-spice-sync-service/SpiceUtils.cs [hunk 2]
```diff
@@ -1433,35 +1402,35 @@ public async Task SendFoundWorkers(PerformContext hangfireContext)
...
+            Log.Logger.Error("End of SendFoundWorkers Job");
             hangfireContext.WriteLine("End of SendFoundWorkers Job");
         }
 
         [DisableConcurrentExecution(timeoutInSeconds: 10 * 60)]
         public async Task SendFoundApplications(PerformContext hangfireContext)
         {
             string[] select = { "adoxio_applicationtypeid" };
-            IList<MicrosoftDynamicsCRMadoxioApplicationtype> selectedAppTypes = _dynamicsClient.Applicationtypes.Get(filter: "adoxio_requiressecurityscreening eq true", select: select).Value;
+            IList<MicrosoftDynamicsCRMadoxioApplicationtype> selectedAppTypes = _dynamicsClient.Applicationtypes.Get(filter: "adoxio_haslesection eq true", select: select).Value;
```
Comment (asanchezr) [inline carla-spice-sync-service/SpiceUtils.cs:1418]: I guess the field got changed in Dynamics? It used to be `adoxio_requiressecurityscreening`. Not sure how it relates to `haslesection`...

File: carla-spice-sync-service/Validation.cs [hunk 3]
```diff
@@ -2,7 +2,7 @@
 using System.Text.RegularExpressions;
 using Gov.Lclb.Cllb.Interfaces;
 using Gov.Lclb.Cllb.Interfaces.Spice.Models;
-using Microsoft.Extensions.Logging;
+using Serilog;
```
Comment (asanchezr) [inline carla-spice-sync-service/Validation.cs:5]: I like the usage of a better logger in this service and the general code cleanup

File: cllc-interfaces/SPICE/SpiceClient.cs [hunk 4]
```diff
@@ -23,6 +26,35 @@ public partial class SpiceClient : ServiceClient<SpiceClient>, ISpiceClient
...
+            if (!string.IsNullOrEmpty(spiceUri) && !string.IsNullOrEmpty(token))
+            {
+                BaseUri = new Uri(spiceUri);
+                string bearer_token = $"Bearer {token}";
+
+                _client.DefaultRequestHeaders.Add("Accept", "application/json");
+                _client.DefaultRequestHeaders.Add("Authorization", bearer_token);
+            }
+
+        }
```
Comment (asanchezr) [inline cllc-interfaces/SPICE/SpiceClient.cs:57]: Really like this refactor - way cleaner!

File: cllc-public-app/ClientApp/src/app/shared/components/related-licence-picker/related-licence-picker.component.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+          this.licenceRequestInProgress = true;
+        }),
+        switchMap(value => this.licenceDataService.getAutocomplete(value))
+      )
+      .subscribe(data => {        
+        this.autocompleteLicences = data;
+        this.licenceRequestInProgress = false;
+        this.inputAutoComplete.openPanel();
+      });
+  }
```
Comment (asanchezr) [inline cllc-public-app/ClientApp/src/app/shared/components/related-licence-picker/related-licence-picker.component.ts:45]: This looks great. I'd suggest looking into `debounce()` - [link](https://rxjs-dev.firebaseapp.com/api/operators/debounce) to help with performance. No need to fire off API requests as the user is typing into the auto-complete box. We can discuss further tomorrow the way I've seen it done in other projects.

File: openshift/clean.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+'use strict';
+
+const fs = require('fs');
+
+let rawdata = fs.readFileSync('student.json');
+let student = JSON.parse(rawdata);
+
+var args = process.argv.slice(2);
```
Comment (asanchezr) [inline openshift/clean.js:8]: Is this file supposed to be in the repo? I can't find `student.json`...

**Review by asanchezr — COMMENTED** [review]
Lot's of good cleanup. I have some comments that would like to discuss

File: carla-spice-sync-service/SpiceUtils.cs (hunk 2 above)
Comment (tgunderson) [reply carla-spice-sync-service/SpiceUtils.cs:1418]: Yes, unfortunately I didn't realize this flag was being used by the service; we use it to toggle some items in the front end and part of the migration to the new flow involved disabling the adoxio_requiressecurityscreening flag.  adoxio_haslesection, when combined with the filter to look at the Sent to SPD flag, will have the correct result though.

File: cllc-public-app/Controllers/LicensesController.cs [hunk 7]
```diff
@@ -690,6 +785,23 @@ private MicrosoftDynamicsCRMadoxioApplication CreateApplication(string licenceId
...
+                var relatedLicence = _dynamicsClient.GetLicenceByIdWithChildren(relatedLicenceId);
+                // set the establishment address to be that of the related licence.
+                application.AdoxioEstablishmentaddressstreet = relatedLicence.AdoxioEstablishment.AdoxioAddressstreet;
+                application.AdoxioEstablishmentaddresscity = relatedLicence.AdoxioEstablishment.AdoxioAddresscity;
+                application.AdoxioEstablishmentaddresspostalcode = relatedLicence.AdoxioEstablishment.AdoxioAddresspostalcode;
+
+                // TODO - the following fields do not appear to be in Dynamics yet
+
+                // Assigned Licensee == Licensee of selected licence
+                // Related Licence == Licence lookup of the manufacturers' licence
```
Comment (tgunderson) [inline cllc-public-app/Controllers/LicensesController.cs:801]: Assigned Licensee is just adoxio_applicant
Created ticket to have Related Licence added (we had Related Application, which I had it confused with) https://justice.gov.bc.ca/jira/browse/LCSD-4926

File: openshift/clean.js (hunk 6 above)
Comment (GeorgeWalker) [reply openshift/clean.js:8]: Good catch, that's some sample code that wasn't supposed to be there.

File: cllc-public-app/ClientApp/src/app/shared/components/related-licence-picker/related-licence-picker.component.ts (hunk 5 above)
Comment (GeorgeWalker) [reply cllc-public-app/ClientApp/src/app/shared/components/related-licence-picker/related-licence-picker.component.ts:45]: Good idea.  I suggest we save that for a future update though

File: cllc-public-app/ViewModels/ApplicationTypeNames.cs (hunk 1 above)
Comment (GeorgeWalker) [reply cllc-public-app/ViewModels/ApplicationTypeNames.cs:10]: good catch, fixed in updated.

**Pull request merged**